plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/config.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Union
+
+from hydra.core.config_store import ConfigStore
+from hydra.types import ObjectConf
+
+
+@dataclass
+class NevergradOptimConf:
+
+    # name of the Nevergrad optimizer to use. Here is a sample:
+    #   - "OnePlusOne" extremely simple and robust, especially at low budget, but
+    #     tends to converge early.
+    #   - "CMA" very good algorithm, but may require a significant budget (> 120)
+    #   - "TwoPointsDE": an algorithm good in a wide range of settings, for significant
+    #     budgets (> 120).
+    #   - "Shiva" an algorithm aiming at identifying the best optimizer given your input
+    #     definition (work in progress, it may still be ill-suited for low budget)
+    # find out more within nevergrad's documentation:
+    # https://github.com/facebookresearch/nevergrad/
+    optimizer: str = "OnePlusOne"
+
+    # total number of function evaluations to perform
+    budget: int = 80
+
+    # number of parallel workers for performing function evaluations
+    num_workers: int = 10
+
+    # set to true if the function evaluations are noisy
+    noisy: bool = False
+
+    # set to true for performing maximization instead of minimization
+    maximize: bool = False
+
+    # optimization seed, for reproducibility
+    seed: Optional[int] = None
+
+
+@dataclass
+class ScalarSpec:
+
+    # lower bound if any
+    lower: Optional[float] = None
+
+    # upper bound if any
+    upper: Optional[float] = None
+
+    # initial value
+    # default to the middle point if completely bounded
+    init: Optional[float] = None
+
+    # step size for an update
+    # defaults to 1 if unbounded
+    # or 1/6 of the range if completely bounderd
+    step: Optional[float] = None
+
+    # cast to integer
+    integer: bool = False
+
+    # logarithmically distributed
+    log: bool = False
+
+
+@dataclass
+class NevergradFullConf:
+
+    # configuration of the optimizer
+    optim: NevergradOptimConf = NevergradOptimConf()
+
+    # default parametrization of the search space
+    # can be specified:
+    # - as a string, like commandline arguments
+    # - as a list, for categorical variables
+    # - as a full scalar specification
+    parametrization: Optional[Dict[str, Union[List[Any], str, ScalarSpec]]] = None
```

Comment:
This is how I interpreted your comment on the other diff:
https://github.com/facebookresearch/hydra/pull/531#pullrequestreview-393145289

One could go one step further by having the scalar specification here. But it seems it does not work at all (probably because of the Union... the spec is a bit errr messy)

Reply:
> > not at all. what version of OmegaConf are you on?
> 
> 2.0.0rc21

Can you try to create a minimal repro case I can play with that shows that bad error?

Reply:
> This is how I interpreted your comment on the other diff:
> [#531 (review)](https://github.com/facebookresearch/hydra/pull/531#pullrequestreview-393145289)
> 
> One could go one step further by having the scalar specification here. But it seems it does not work at all (probably because of the Union... the spec is a bit errr messy)
> 
> Let me know if there's a way to fix this, or I'll revert to a mere `Any`

Yeah, unions are a no-go for now.
you can put Any and do the checks in runtime.

What I mean is that can have shared logic deciding the priority between the parametrizaton and the command line (how what type are we ending up with and all that).

Reply:
"minimal" probably not, I'm lost with the config.
All I know is that on this branch, if you run `python example/dummy_training.py -m` from within the nevergrad plugin directory, you get:
`omegaconf.errors.ValidationError: Unsupported value type : typing.Union[typing.List[typing.Any], str, hydra_plugins.hydra_nevergrad_sweeper.config.ScalarSpec]`
If I remove the Union, I get yet a different type of error, which I figured was probably related to the fact that the typing was wrong then, but it was not very explicit

Reply:
Okay, I will take a look later.
can you paste the whole stack trace?

Reply:
The thread confuses me, there are 2 conversations in one

So I think I got it wrong: the repro was for the Union problem, which as you mentioned is not supported. That can be really messy so that seems logical, I revert back to Any

I wont have time either for the unclear error message though, but the way I did it was:
- modify `num_workers=10` to `num_workers=blublu` in 'example/config.yaml'
- run `python example/dummy_training.py -m`

Then I got:

```
Traceback (most recent call last):
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/nodes.py", line 203, in validate_and_convert
    val = int(value)
ValueError: invalid literal for int() with base 10: 'blublu'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 283, in merge_with
    BaseContainer._map_merge(self, other)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 257, in _map_merge
    dest_node._set_value(src_value)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/nodes.py", line 39, in _set_value
    self._val = self.validate_and_convert(value)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/nodes.py", line 207, in validate_and_convert
    raise ValidationError("Value '$VALUE' could not be converted to Integer")
omegaconf.errors.ValidationError: Value '$VALUE' could not be converted to Integer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "example/dummy_training.py", line 29, in <module>
    dummy_training()
  File "/Users/jrapin/fairinternal/hydra/hydra/main.py", line 31, in decorated_main
    strict=strict,
  File "/Users/jrapin/fairinternal/hydra/hydra/_internal/utils.py", line 235, in run_hydra
    overrides=args.overrides,
  File "/Users/jrapin/fairinternal/hydra/hydra/_internal/hydra.py", line 118, in multirun
    with_log_configuration=True,
  File "/Users/jrapin/fairinternal/hydra/hydra/_internal/hydra.py", line 447, in compose_config
    config_name=config_name, overrides=overrides, strict=strict
  File "/Users/jrapin/fairinternal/hydra/hydra/_internal/config_loader_impl.py", line 96, in load_configuration
    hydra_cfg, job_cfg, job_cfg_load_trace, defaults, split_at
  File "/Users/jrapin/fairinternal/hydra/hydra/_internal/config_loader_impl.py", line 397, in _merge_defaults
    hydra_cfg = merge_defaults(hydra_cfg, system_list)
  File "/Users/jrapin/fairinternal/hydra/hydra/_internal/config_loader_impl.py", line 364, in merge_defaults
    merged_cfg.merge_with(job_cfg)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 296, in merge_with
    self._format_and_raise(key=None, value=None, cause=e)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/base.py", line 96, in _format_and_raise
    type_override=type_override,
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/_utils.py", line 427, in format_and_raise
    raise cause
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 283, in merge_with
    BaseContainer._map_merge(self, other)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 249, in _map_merge
    dest_node.merge_with(src_value)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 296, in merge_with
    self._format_and_raise(key=None, value=None, cause=e)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/base.py", line 96, in _format_and_raise
    type_override=type_override,
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/_utils.py", line 427, in format_and_raise
    raise cause
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 283, in merge_with
    BaseContainer._map_merge(self, other)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 249, in _map_merge
    dest_node.merge_with(src_value)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 296, in merge_with
    self._format_and_raise(key=None, value=None, cause=e)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/base.py", line 96, in _format_and_raise
    type_override=type_override,
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/_utils.py", line 427, in format_and_raise
    raise cause
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 283, in merge_with
    BaseContainer._map_merge(self, other)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 249, in _map_merge
    dest_node.merge_with(src_value)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 296, in merge_with
    self._format_and_raise(key=None, value=None, cause=e)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/base.py", line 96, in _format_and_raise
    type_override=type_override,
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/_utils.py", line 427, in format_and_raise
    raise cause
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 283, in merge_with
    BaseContainer._map_merge(self, other)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 249, in _map_merge
    dest_node.merge_with(src_value)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 296, in merge_with
    self._format_and_raise(key=None, value=None, cause=e)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/base.py", line 96, in _format_and_raise
    type_override=type_override,
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/_utils.py", line 499, in format_and_raise
    raise ex from cause
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 283, in merge_with
    BaseContainer._map_merge(self, other)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/basecontainer.py", line 257, in _map_merge
    dest_node._set_value(src_value)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/nodes.py", line 39, in _set_value
    self._val = self.validate_and_convert(value)
  File "/Users/jrapin/.venvs/hydraenv/lib/python3.7/site-packages/omegaconf/nodes.py", line 207, in validate_and_convert
    raise ValidationError("Value '$VALUE' could not be converted to Integer")
omegaconf.errors.ValidationError: Value 'None' could not be converted to Integer
	full_key:
	reference_type=NevergradOptimConf
	object_type=NevergradOptimConf
```

Reply:
> The thread confuses me, there are 2 conversations in one
> 
> So I think I got it wrong: the repro was for the Union problem, which as you mentioned is not supported. That can be really messy so that seems logical, I revert back to Any
> 
> I wont have time either for the unclear error message though, but the way I did it was:

The stacktrace is enough, thanks.

Reply:
About behavior I would have expected but does not seem to work:
If I add non-existing parameters to the optimizer config for the example (eg adding `blublu=10` right after `num_workers=10` in the example config.yaml) then it is just ignored.
Is that expected? I would have thought it would either raise an exception, or at least a warning.

Apart from that, all seems to work, and I've moved the scalar spec to the other PR.

Reply:
Please be more specific.
You added blublu=10 where?
command line?
the dataclass?
a yaml config file?

Reply:
ah, in config.yaml. looking.

Reply:
Adding a key to optim that does not exist in the corresponding NevergradOptimConf should definitely result in an error.

Reply:
I will look into both issues. thanks.